pkg/flow/internal/controller/loader.go

```diff
@@ -182,18 +158,53 @@ func (l *Loader) Apply(parentScope *vm.Scope, blocks []*ast.BlockStmt, configBlo
 	l.components = components
 	l.graph = &newGraph
 	l.cache.SyncIDs(componentIDs)
-	l.blocks = blocks
+	l.blocks = componentNodeBlocks
 	l.cm.componentEvaluationTime.Observe(time.Since(start).Seconds())
 	return diags
 }
 
-func (l *Loader) populateGraph(g *dag.Graph, blocks []*ast.BlockStmt) diag.Diagnostics {
+func (l *Loader) loadNewGraph(parentScope *vm.Scope, componentNodeBlocks []*ast.BlockStmt, configBlocks []*ast.BlockStmt, globals ComponentGlobals, isInModule bool) (dag.Graph, diag.Diagnostics) {
+	var g dag.Graph
+	// Fill our graph with config blocks.
+	diags := l.populateConfigBlockNodes(&g, configBlocks, globals, isInModule)
+
 	// Fill our graph with components.
+	componentNodeDiags := l.populateComponentNodes(&g, componentNodeBlocks)
+	diags = append(diags, componentNodeDiags...)
+
+	// Write up the edges of the graph
+	wireDiags := l.wireGraphEdges(parentScope, &g)
+	diags = append(diags, wireDiags...)
+
+	// Validate graph to detect cycles
+	err := dag.Validate(&g)
+	if err != nil {
+		diags = append(diags, multierrToDiags(err)...)
+		return g, diags
+	}
+
+	// Copy the original graph, this is so we can have access to the original graph for things like displaying a UI or
+	// debug information.
+	l.originalGraph = g.Clone()
+	// Perform a transitive reduction of the graph to clean it up.
+	dag.Reduce(&g)
+
+	return g, diags
+}
+
+func (l *Loader) populateConfigBlockNodes(g *dag.Graph, configBlocks []*ast.BlockStmt, globals ComponentGlobals, isInModule bool) diag.Diagnostics {
+	c, diags := NewConfigNode(configBlocks, l.globals, l.isModule())
```

Comment:
This is where we would loop on the configBlocks instead of inside NewConfigNode. NewConfigNode would only take in a single configBlock instead of the whole slice.